man/erf.1
.Dd October 20, 2019
.Dt ERF 1
.Os
.Sh NAME
.Nm erf
.Nd BioWare ERF (.erf, .mod, .nwm, .sav) archive packer
.Sh SYNOPSIS
.Nm erf
.Op Ar options
.Ar output_archive
.Op Ar
.Sh DESCRIPTION
.Nm
packs files together into a BioWare ERF archive, found in many
BioWare Games as files with the extension .erf, .mod, .nwm or .sav.
Moreover, in some games, a .rim file might be an ERF instead of a RIM.
.Pp
There's several different versions of ERFs.
This tool supports version V1.0, as used by
.Em Neverwinter Nights ,
.Em Knights of the Old Republic ,
.Em Knights of the Old Republic II ,
.Em Jade Empire
and
.Em The Witcher
and version V2.0 as used by
.Em Dragon Age: Origins
and version V2.2 as used by
.Em Dragon Age: Origins
and
.Em Dragon Age 2
.Pp
Unsupported Features:
.Bl -bullet -compact
.It
Generation of archives with stripped filenames.
.It
Generation of Blowfish-encrypted archives.
.It
Generation of XOR-encrypted archives.
.It
Setting of a description for archives.
.It
Usage of paths in archives.
.El
.Sh OPTIONS
.Bl -tag -width xxxx -compact
.It Fl h
.It Fl Fl help
Show a help text and exit.
.It Fl Fl version
Show version information and exit.
.It Fl Fl erf
Set archive ID to ERF.
.It Fl Fl mod
Set archive ID to MOD.
.It Fl Fl hak
Set archive ID to HAK.
.It Fl Fl sav
Set archive ID to SAV.
.It Fl Fl v10
Generate a V1.0 ERF file
.It Fl Fl v20
Generate a V2.0 ERF file
.It Fl Fl v22
Generate a V2.2 ERF file
.It Fl Fl bzlib
Compress using BioWare zlib method
.It Fl Fl zlib
Compress using headerless zlib method
.It Fl Fl jade
Unalias file types according to
.Em Jade Empire
rules.
.Pp
.Em Jade Empire
reuses a few file extension IDs differently than other BioWare games.
To correctly write Jade Empire ERF archives, use this flag.
.It Ar output_archive
The ERF archive to be written.
.It Ar files
One or more files to pack together.
.El
.Sh EXAMPLES
Pack some files together into an ERF archive:
.Pp
.Dl $ erf archive.erf file1.dat file2.dat file3.dat
.Pp
Pack some files together into a SAV archive:
.Pp
.Dl $ erf --sav archive.sav file1.dat file2.dat file3.dat
.Pp
Pack some files together into a V2.0 archive:
.Pp
.Dl $ erf --v20 archive.sav file1.dat file2.dat file3.dat
.Pp
Pack some files together into a V2.2 archive with headerless zlib compression:
.Pp
.Dl $ erf --v22 --zlib archive.sav file1.dat file2.dat file3.dat
.Sh SEE ALSO
.Xr unerf 1 ,
.Xr unherf 1
.Pp
More information about the xoreos project can be found on
.Lk https://xoreos.org/ "its website"
.Ns .
.Sh AUTHORS
This program is part of the xoreos-tools package, which in turn is
part of the xoreos project, and was written by the xoreos team.
Please see the
.Pa AUTHORS
file for details.


NEWS
Changes in xoreos-tools version 0.0.2
=====================================

This is the first official release of xoreos-tools, together with xoreos.

Included are:

- Tools for handling resource archives
  - unkeybif
  - unerf
  - unrim
  - unherf
  - unnds
  - desmall
- Tools converting basic resource formats into human-readable form
  - gff2xml
  - tlk2xl
  - convert2da
- Tools converting graphics formats into TGA
  - xoreostex2tga
  - unnsbtx
  - nbfs2tga
  - ncgr2tga
  - cbgt2tga
  - cdpth2tga


src/erf.cpp
/* xoreos-tools - Tools to help with xoreos development
 *
 * xoreos-tools is the legal property of its developers, whose names
 * can be found in the AUTHORS file distributed with this source
 * distribution.
 *
 * xoreos-tools is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 3
 * of the License, or (at your option) any later version.
 *
 * xoreos-tools is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with xoreos-tools. If not, see <http://www.gnu.org/licenses/>.
 */

/** @file
 *  Tool to pack ERF (.erf, .mod, .nwm, .sav) archives.
 */

#include <set>

#include "src/common/error.h"
#include "src/common/platform.h"
#include "src/common/cli.h"
#include "src/common/readfile.h"
#include "src/common/writefile.h"
#include "src/common/filepath.h"

#include "src/aurora/erfwriter.h"
#include "src/aurora/util.h"

#include "src/util.h"

static const uint32 kERFID = MKTAG('E', 'R', 'F', ' ');
static const uint32 kMODID = MKTAG('M', 'O', 'D', ' ');
static const uint32 kHAKID = MKTAG('H', 'A', 'K', ' ');
static const uint32 kSAVID = MKTAG('S', 'A', 'V', ' ');

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &archive, std::set<Common::UString> &files,
                      Aurora::ERFWriter::Version &version, Aurora::ERFWriter::Compression &compression,
                      uint32 id, Aurora::GameID &game);

int main(int argc, char **argv) {
	initPlatform();

	try {
		std::vector<Common::UString> args;
		Common::Platform::getParameters(argc, argv, args);

		Aurora::GameID game = Aurora::kGameIDUnknown;

		int returnValue = 1;
		uint32 id = kERFID;
		Common::UString archive;
		Aurora::ERFWriter::Version version = Aurora::ERFWriter::kERFVersion10;
		Aurora::ERFWriter::Compression compression = Aurora::ERFWriter::kCompressionNone;
		std::set<Common::UString> files;

		if (!parseCommandLine(args, returnValue, archive, files, version, compression, id, game))
			return returnValue;

		if (compression != Aurora::ERFWriter::kCompressionNone && version != Aurora::ERFWriter::kERFVersion22)
			throw Common::Exception("Compression is only allowed in ERF V2.2");

		Common::WriteFile writeFile(archive);

		size_t i = 1;
		Aurora::ERFWriter erfWriter(id, files.size(), writeFile, version, compression);
		for (std::set<Common::UString>::const_iterator iter = files.begin(); iter != files.end(); ++iter, ++i) {
			std::printf("Packing %u/%u: %s ... ", (uint)i, (uint)files.size(), iter->c_str());
			std::fflush(stdout);

			Common::UString file = *iter;
			Common::ReadFile fileStream(file);

			const Aurora::FileType type = TypeMan.unaliasFileType(TypeMan.getFileType(file), game);

			erfWriter.add(Common::FilePath::getStem(file), type, fileStream);
			std::printf("Done\n");
		}
	} catch (...) {
		Common::exceptionDispatcherError();
	}

	return 0;
}

bool parseCommandLine(const std::vector<Common::UString> &argv, int &returnValue,
                      Common::UString &archive, std::set<Common::UString> &files,
                      Aurora::ERFWriter::Version &version, Aurora::ERFWriter::Compression &compression,
                      uint32 id, Aurora::GameID &game) {
	using Common::CLI::NoOption;
	using Common::CLI::kContinueParsing;
	using Common::CLI::Parser;
	using Common::CLI::Callback;
	using Common::CLI::ValGetter;
	using Common::CLI::ValAssigner;
	using Common::CLI::makeEndArgs;
	using Common::CLI::makeAssigners;
	using Aurora::GameID;

	NoOption archiveOpt(false, new ValGetter<Common::UString &>(archive, "output archive"));
	NoOption filesOpt(true, new ValGetter<std::set<Common::UString> &>(files, "files[...]"));

	Parser parser(argv[0], "BioWare ERF archive packer",
	              "",
	              returnValue,
	              makeEndArgs(&archiveOpt, &filesOpt));

	parser.addSpace();
	parser.addOption("erf", "Set ERF as archive id (default)",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<uint32>(kERFID, id)));
	parser.addOption("mod", "Set MOD as archive id",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<uint32>(kMODID, id)));
	parser.addOption("hak", "Set HAK as archive id",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<uint32>(kHAKID, id)));
	parser.addOption("sav", "Set SAV as archive id",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<uint32>(kSAVID, id)));
	parser.addSpace();
	parser.addOption("v10", "Generate a V1.0 ERF file (default)",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Aurora::ERFWriter::Version>(Aurora::ERFWriter::kERFVersion10, version)));
	parser.addOption("v20", "Generate a V2.0 ERF file",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Aurora::ERFWriter::Version>(Aurora::ERFWriter::kERFVersion20, version)));
	parser.addOption("v22", "Generate a V2.2 ERF file",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Aurora::ERFWriter::Version>(Aurora::ERFWriter::kERFVersion22, version)));
	parser.addSpace();
	parser.addOption("bzlib", "Compress using BioWare zlib method",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Aurora::ERFWriter::Compression>(Aurora::ERFWriter::kCompressionBiowareZlib, compression)));
	parser.addOption("zlib", "Compress using headerless zlib method",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<Aurora::ERFWriter::Compression>(Aurora::ERFWriter::kCompressionHeaderlessZlib, compression)));
	parser.addSpace();
	parser.addOption("jade", "Unalias file types according to Jade Empire rules",
	                 kContinueParsing,
	                 makeAssigners(new ValAssigner<GameID>(Aurora::kGameIDJade, game)));

	return parser.process(argv);
}
